lib/doorkeeper/oauth/authorization_code_request.rb

```diff
@@ -93,7 +84,7 @@ def validate_redirect_uri
       # if either side (server or client) request PKCE, check the verifier
       # against the DB - if PKCE is supported
       def validate_code_verifier
-        return true unless grant.uses_pkce? || code_verifier
+        return true unless grant.uses_pkce?
```

Comment:
This check had to be removed as well, and caused the changed spec in `spec/requests/flows/authorization_code_spec.rb` to fail.

Reply:
I think we can thinking about `validate_code_verifier` as following:
- `pkce non supported` (pkce_supported? == false) -> return true

- `pkce is supported` (pkce_supported? == true)

||code_challenge.present?|code_challenge.blank?|
|---|---|---|
|code_verifier.present?| (1) run validate | (2) run validate |
|code_verifier.blank?| (4) (already validated at validate_params)return false | (3) return true |

So, as the table, currently changes ([Line 87 and 88](https://github.com/doorkeeper-gem/doorkeeper/pull/1415/files#diff-5c1f9daf0b81d01ceeeb79eaea1cbbe2R87-R88)) will not run validated on (2), when code_verifier is present, and code_challenge is not sent from client at authorization request, this can causing [Downgrade Attack on PKCE](https://danielfett.de/2020/05/16/pkce-vs-nonce-equivalent-or-not/#downgrade-attack-on-pkce)

So I think this validation can write as below:

```ruby
return true unless pkce_supported?
return false if grant.code_challenge.present? && code_verifier.blank? # (*)
# (*) is currently validated at `validate_params`, but I still wrote it because
# I think we should consider this is independent from `validate_params`, 
# any change to `validate_params` should not break `validate_code_verifier` 
# (cause `validate_params` is to informs that client missing necessary parameter)

if grant.code_challenge_method == "S256"
.....
```

Reply:
@linhdangduy thanks, this makes sense to me. :+1: 

I updated the method now as per your suggestion, though I simplified the 4/5 case into `return grant.code_challenge.blank? if code_verifier.blank?` and added a spec for the case where PKCE is enabled but both `code_challenge` and `code_verifier` are blank. I think the remaining cases should already be covered by the specs in `spec/lib/oauth/authorization_code_request_spec.rb`.